File: leda/R02-01/R02-01-Rot-RecursiveSorting-Quick3-Merge-environment/src/main/java/sorting/divideAndConquer/InsertionSort.java
package sorting.divideAndConquer;

import sorting.AbstractSorting;
import util.Util;

public class InsertionSort<T extends Comparable<T>> extends AbstractSorting<T> {

    private void insertionSort(T[] array, int start, int end) {
        for (int i = start + 1; i < end + 1; i++) {
            for (int j = i; j > start; j--) {
                boolean isCurrentSmallerThanPrevious = array[j].compareTo(array[j - 1]) < 0;
                if (isCurrentSmallerThanPrevious) {
                    Util.swap(array, j - 1, j);
                }
            }
        }
    }

    public void sort(T[] array, int leftIndex, int rightIndex) {
        insertionSort(array, leftIndex, rightIndex);
    }
}

File: eda/algorhitms/Python/insertionSort/insertionSort.py
def insertionSort(array, start, end):
    for i in range(start, end + 1):
        for j in range(i, start, -1):
            if array[j] <= array[j - 1]:
                swap(array, j, j - 1)
            else:
                break


def swap(array, id1, id2):
    temp = array[id1]
    array[id1] = array[id2]
    array[id2] = temp


array = [9, 3, 7, 99, 5, 2]
insertionSort(array, 0, len(array) - 1)
print(array)
